src/apps/LISS/seedport/seedport.1
.TH seedport 1 "13 January 1998"
.IX seedport
.SH NAME
seedport - connect to a LISS1 or LISS2 server and output SEED to stdout
.SH SYNOPSIS
.B seedport
.B -x
.B -v
.B -b 
.IR size
.B -s
.IR selectfile
.B -p
.IR tcp port
.B -L
.B -c
.IR mask
.B -f 
.IR secs
.B -i
.IR sequence
.B -P
.IR pidfile
.I host
.SH DESCRIPTION
Connect to a 
.IR LISS-1
(netport) or a 
.IR LISS-2
(extended mode) port and pull a real-time data stream from it.
.SH OPTIONS
.PP
The different flags specify the method for connecting to a port.
.TP
.B -x
Output data to the standard output in LISS-2 (extended mode).
Extended mode is comprised simply of SEED records enclosed
in a representation which automatically specifies the source, format,
and size of the data.  If this flag is not used, data will be
send to standard output in raw SEED.
.TP
.B -v
Emit diagnostic messages to standard error.  Otherwise 
.IR seedport
will be quiet.
.TP
.B -L 
Connect into the LISS-2 network automatically.  There is no
need with this option to specify host names or port numbers.
.IR Seedport
will query information servers on the internet to service
your request.
.TP
.B -b
.IR size
.br
Override the default record sizes.  This corresponds with the
SEED
.I logical record size.
The default record size for files is 512.
.TP
.B -p
.IR port
.br
The port number on the LISS-1 (or LISS-2) compatible server that 
.IR seedport
will attach to.  This port is a TCP socket.  The default
if unspecified is 4000.
.TP
.B -s
.IR selectfile
.br
Allows complex selection to be performed on the input data
stream.  Combinations of Network, Station, Location, Channel
and time-range can be used.  Please see the 
.BR selectfile (4)
manual page for information on how to construct a select file.
If 
.IR seedport
is connecting to a bi-directional LISS-2 port, the filter will
be sent up the server.  This can greatly lower the network load
by transmitting only the desired data, instead of sending it
all and throwing away undesired data at the client end.
.TP
.B -E
Ignore End of Files, or network errors on the port.  Will
automatically reconnect if the port is dropped.  Note that if
this option is set, the program will never exit.
.TP
.B -c
.IR filemask
.br
Instead of sending the data received to the standard output,
and series of files will be created with the name specified
in the
.IR filemask .
The mask must contain the %d (C printf format) characters, such 
as data%d or data%04d (must be able to hold a 4 digit number).
These will be substituted with the file number.  A new file
will be created with the next sequence number each time the
timer expires.  The file that is current being written will
have the string 
.IR _live
appended to it.
.TP
.B -f
.IR seconds
.br
Flush and start a new capture file (-c above) after this number
of seconds (the timer value).  Defaults to 3600 (1 hour).
.TP
.B -i
.IR sequence
.br
Set the file number of the initial capture file to this number.
The default is 1.
.TP
.B -P
.IR pidfile
.br
Keep a watchdog file.
When selected, a filename of
.RI /tmp/ pidfile .tmp
(where pidfile is
the value specified here) is written out every time a record
is processed.
There may be some pathological socket I/O
situations that
.B seedport
might not be able to recover from (hopefully we've
found and fixed most of these).  A separate watchdog
process can keep an eye on a 
.B seedport
process via this file and kill it and restart it if it
becomes hung.  The pidfile contains the pid of the process
and the epochal time when a record was last processed.  An
initial entry is placed in the file when the program is
started.  This will keep the watchdog from killing the
program before any input comes in.
.PP
The argument is either the host name (or IP address) of the LISS-1
or LISS-2 server.
.SH EXAMPLES
Get data from the COLA LISS-1 server:
.RS
.B "example% seedport -p 4060 asl6.cr.usgs.gov"
.RE
.SH HISTORY
.IR seedport
was extracted from 
.BR dumpseed(1)
to form a standalone program to permit
users to receive raw seed from a LISS.
.SH "SEE ALSO"
.BR dumpseed(1) ,
.BR getpid(2) ,
.BR time(2) ,
.BR selectfile(4) ,
.BR SEED(4) 
.SH AUTHOR
Public Domain Software by Scott Halbert - Allied Signal Technical
Services Corporation under contract to the Albuquerque Seismological
Laboratory - United States Geological Survey - Department of Interior,
United States, North American Continent, Earth/Sol 3, Sagitarius Arm of
Milky Way Galaxy.


CHANGELOG
21-Jul-2003     GTM     Version 2.0.0
                        Package now consists of dumpseed, seedport, seedfilter and seedforker

28-Oct-2002     GTM     Version 1.01
                        Altered the following to implement data quality indicator flags of D, Q or R
				combiner:  combiner.c
				cutter:    steimdump.c
				dumpseed:  dumpseed.c, prtseedbdf.c, prtseedrec.c, steimdump.c
				seedspans: seedspans.c
                        Corrected and enhanced Usage for dumpseed

??-Sep-97	SH	RELEASE 0.0

16-Sep-97	SH	Port non-oracle parts of DCC libraries to
			portable library.  Redo all makefiles and
			configuration to GNU autoconf

[Beginning of Project]

src/apps/LISS/seedport/seedport.c
#include <dcc_std.h>
#include <dcc_time.h>
#include <seed_data.h>
#include <seed_data_proto.h>
#include <dcc_misc.h>
#include <dcc_seed_select.h>

#include "seedport_proto.h"

/*
 * client program to catch data from dpnetport
 */

#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <sys/uio.h>

time_t now;

int extendedoutput=0;
int verbose=0;
int liss2=0;
int ignoreeof=0;
int blocksize=0;
int tel_port=4000;
char *select_file=NULL;
char *pidfile=NULL;
char *capture=NULL;
int flushint=3600;
int startint=1;

struct sel3_root dumproot;

#define MAX_FILES 10
char *ifil[MAX_FILES];
int num_files=0;

int seed_size = 512;  /* OK default for now, since we're just
			 sending all data to stdout -- at this
			 point we don't actually care how big it
			 is - later, with extendedoutput format, we will */

unsigned char buf[8192];
int bufpop=0,jettison=0;

char outf_name[500];
char outf_dest[500];
FILE *outf=NULL;

void process_record(char *inseed)
{

  int i;
  time_t curr;
  FILE *outputfile;
  ITEMLIST *taglist=NULL;

  outputfile = stdout;

  if (pidfile)
    watchdog(pidfile);

  if (select_file) {
    int interesting;
    char n[10],s[10],l[10],c[10];

    SH_Get_Idents((SEED_DATA *) inseed,n,s,l,c);
    TrimString(n);
    Upcase(n);
    TrimString(s);
    Upcase(s);
    TrimString(l);
    Upcase(l);
    TrimString(c);
    Upcase(c);

    interesting = SelectDateInteresting(&dumproot,
					SH_Start_Time((SEED_DATA *) inseed),
					SH_End_Time((SEED_DATA *) inseed),
					n,s,l,c,
					&taglist);

    if (!interesting) return;

  }

  /* Are we capturing? */

  if (capture) {

    /* See if flush interval is over - if so, we start new file */

    curr = time(NULL);
    if (curr>now+flushint) {

      /* Close old file and rename it - if any */

      if (outf) {
	fclose(outf);
	
	/* mv temp name to permanent name */

	unlink(outf_dest);
	i = link(outf_name, outf_dest);
	if (i!=0) 
	  fprintf(stderr,"Couldn't link new file %s - %s\n",
		  outf_dest, strerror(errno));
	else
	  unlink(outf_name);

	outf = NULL;
      }

      /* Open new file */

      sprintf(outf_dest, capture, startint++);
      strcpy(outf_name, outf_dest);
      strcat(outf_name, "_live");

      outf = fopen(outf_name,"w");
      if (outf==NULL) {
	printf("Couldn't write file %s\n",outf_name);
	exit(20);
      }
      if (verbose) 
	fprintf(stderr,"[Open capture file %s]\n",
		outf_dest);

    now = curr; /* When did we open the file? */

    }

    outputfile = outf;

  }

  if (!extendedoutput) {
    i = write(fileno(outputfile), inseed, seed_size);
    if (i!=seed_size) {
      fprintf(stderr,"Cannot write to output\n");
      exit(30);
    }
    fflush(outputfile);
    return;
  }

}

_SUB int process_telemetry(char *host, int port)
{
  struct sockaddr_in sin;
  struct hostent *hp, *gethostbyname();
  int fd, count, optval, i;

  if (verbose)
    fprintf(stderr,"[Connecting to remote host: %s (port %d) blocksize %d]\n",
	    host,port,seed_size);

  /*
   * Translate the hostname into an internet address
   */
  hp = gethostbyname(host);
  if(hp == 0) {
    fprintf(stderr, "%s: unknown host\n",host);
    exit(1);
  }
  /*
   * We now create a socket which will be used for our connection.
   */
  fd = socket(AF_INET, SOCK_STREAM, 0);
  if(fd < 0) {
    fprintf(stderr,"Couldn't open socket on client socket - %s\n",
	    strerror(errno));
    return(1);
  }

  /*
   *  Before we attempt a connect, we fill up the "sockaddr_in" structure
   */
  memcpy((char *)&sin.sin_addr, hp->h_addr , hp->h_length);
  sin.sin_family      = AF_INET;
  sin.sin_port        = htons(port);
  /*
   *  Now attempt to connect to the remote host .....
   */
  if(connect(fd, (struct sockaddr *) &sin, sizeof(sin)) < 0) {
    fprintf(stderr,"Couldn't connect to server %s - %s\n",host,
	    strerror(errno));
    return(2);
  }

  if (verbose)
    fprintf(stderr,"Connected to remote host %s port %d (%s)\n",
	    host,port,hp->h_name);
  fflush(stderr);

  /* Turn on keepalive on socket so we won't hang if it gets stuck */

  optval = 1;
  i = setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, (char *) &optval, sizeof(int));
  
  if (i==(-1)) 
    fprintf(stderr,"Couldn't set keepalive on client socket - ignored - %s\n",
	    strerror(errno));

  bufpop = 0;

  for (;;) {

    /*
     * Read from the network
     */
    count = recv(fd, &buf[bufpop], 256, 0);
    if(count < 0) {
      /* Unable to read from the terminal */
      fprintf(stderr,"Read error from server - %s\n",
	      strerror(errno));
      return(3);
    }
    if (count==0) {
      if (verbose)
	fprintf(stderr,"[End of file]\n");
      break;
    }
	
    bufpop += count;

    while (bufpop>=seed_size) {

      process_record((char *) buf);
      bufpop -= seed_size;
      if (bufpop>0)
	memcpy(buf,&buf[seed_size],bufpop);
    }

    /*
     * Continue the loop
     */
  }                  

  close(fd);

  return(0);
}

void usage(int argc,char **argv)
{

  fprintf(stderr,"Usage: %s -x -v -b seedrecsize -s selectfile -L -p x -P pidfile -c mask -f secs -i x host\n",
	 argv[0]);
  fprintf(stderr,"\n");
  fprintf(stderr,"    -x         output in extended (LISS2) mode\n");
  fprintf(stderr,"    -v         emit diagnostic messages\n");
  fprintf(stderr,"    -b recsiz  seed record size - defaults to 512 - ignored in extended mode\n");
  fprintf(stderr,"    -s file    filter selections - sends to LISS if enabled\n");
  fprintf(stderr,"    -p port    port to connect to - defaults to 4000\n");
  fprintf(stderr,"    -E         ignore EOF - keep reconnecting to port\n");
  fprintf(stderr,"    -c mask    capture to a series of files named by mask (out%%d)\n");
  fprintf(stderr,"    -f secs    flush interval of capture in seconds (defaults to 3600)\n");
  fprintf(stderr,"    -i init    Initial capture file # (normally 1)\n");

  fprintf(stderr,"    -P file    write a PIDfile\n\n");

  exit(20);
}

int main(int argc,char **argv)
{

  int c;
  extern char *optarg;
  extern int optind;
  int errflg=0;
  int i;
  char test1[500],test2[500];

  while ((c = getopt(argc, argv, "xvLEb:s:p:c:f:i:P:")) != EOF)
    switch (c) {
    case 'x':
      extendedoutput++;
      break;
    case 'v':
      verbose++;
      break;
    case 'L':
      liss2++;
      break;
    case 'E':
      ignoreeof++;
      break;
    case 'b':
      i = atoi(optarg);
      switch(i) {
      case 256:
      case 512:
      case 4096:
	seed_size = i;
	break;
      default:
	fprintf(stderr,"Block size must be 256, 512, or 4096\n");
	errflg++;
	break;
      }
      break;
    case 'p':
      tel_port = atoi(optarg);
      break;
    case 's':
      select_file = strdup(optarg);
      break;
    case 'P':
      pidfile = strdup(optarg);
      break;
    case 'c':
      capture = strdup(optarg);
      break;
    case 'f':
      flushint = atoi(optarg);
      if (flushint<=5) {
	fprintf(stderr,"Flush interval must be >= 5 seconds\n");
	errflg++;
      }
      break;
    case 'i':
      startint = atoi(optarg);
      if (startint<0) {
	fprintf(stderr,"Starting sequence must be >= 0\n");
	errflg++;
      }
      break;
    case '?':
      errflg++;
    }

  if (capture) {
    sprintf(test1,capture,1);
    sprintf(test2,capture,9999);

    if (strcmp(test1,test2)==0) {
      fprintf(stderr,"You must specify capture masks with the %%d operator - like out%%d.seed\n");
      errflg++;
    }
  }

  if (liss2) {
    printf("LISS2 server connections not supported yet\n");
    errflg++;
  }

  if (extendedoutput) {
    printf("Extended (LISS2) mode not yet supported\n");
    errflg++;
  }

  if (errflg) 
    usage(argc,argv);

  if (pidfile)
    watchdog(pidfile);

  SelectInit(&dumproot);

  if (select_file) {
    SelectLoadup(&dumproot, select_file);
    if (verbose) 
      SelectList(&dumproot);
  }

  for (; optind < argc; optind++) {
    ifil[num_files++] = strdup(argv[optind]);
    if (num_files+1>MAX_FILES) 
      bombout(EXIT_ABORT,"Too many files - only supports %d",MAX_FILES);
  }
  
  if (num_files<=0) usage(argc,argv);
  if (num_files>1) usage(argc,argv);

  now = 0;

  while(1) {
    i = process_telemetry(ifil[0], tel_port);
    if (!ignoreeof) break;
    sleep(20);
  }

  exit(0);

}
